### PR TITLE
Testing for the document handler

### LINKED ISSUE #286: Write unit tests for the document handler
# Summary

Unit tests do not currently exist for the Telegram document handler.

# Background

**Is your improvement related to a problem? Please describe**:

After writing this handler, it is unclear if we have handled all edge cases. We should have unit tests to thoroughly test this functionality.

**Describe the solution you'd like**:

Full set of unit tests to catch as many edge cases as we can.

**Describe alternatives you've considered**:

No alternatives.

# Details

The owner of this issue needs to think about what a user can and cannot do when uploading a file to Telegram. Then, build unit tests around the desired and undesired functionality to make sure error handling exists in the appropriate locations.

# Outcome

Unit testing for the TG document handler. Allows us to feel safe knowing this handler is thoroughly tested.

### PR DESCRIPTION
**Pull request by 10eMyrT**
Good morning! Considering the comments made to issue #289, this PR introduces testing for the document handler. This is a work in-progress, as I believe your comments will contribute to improving the code. Note that all the tests pass when I run `go test ./...` on my local machine. 
Closes #286

### CODE REVIEW

File: internal/handlers/telegram/handler_test.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+	// Assign the MimeType(Optional) to FileID(Required) if it does not exists
+	if updateObj.Message.Document.MimeType == ""{
+		updateObj.Message.Document.MimeType = updateObj.Message.Document.FileID
+	}
+
+	// Change to FileName(Optional) is the Caption(Optional) is not given
+	if updateObj.Message.Caption == ""{
+		updateObj.Message.Document.FileName = "Random Text"
+	}
```
Comment (10eMyrT) [inline internal/handlers/telegram/handler_test.go:33]: I am cautious about Caption and FileName. They are both optional, meaning that a document may not have either a caption or a filename attached. The test might fail in that case. These are the two options we used in the `handler.go` file. Should we add another option or remove both Caption and FileName from how we format the string sent to IRC?

File: internal/handlers/telegram/handler_test.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+func TestDocument(t *testing.T) {
+	correct := "Someone shared a file (Random File) on Telegram with caption: 'Random Caption'."
+	updateObj := &tgbotapi.Update{
+		Message: &tgbotapi.Message{
+			From: &tgbotapi.User{
+				FirstName: "Someone",
+			},
+			Document: &tgbotapi.Document{
+				FileID: "Random File",
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:17]: This file id will actually be the download link. So something like `"https://teleirc.com/file.txt"` as a fake download link would work.

File: internal/handlers/telegram/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestDocument(t *testing.T) {
+	correct := "Someone shared a file (Random File) on Telegram with caption: 'Random Caption'."
+	updateObj := &tgbotapi.Update{
+		Message: &tgbotapi.Message{
+			From: &tgbotapi.User{
+				FirstName: "Someone",
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:14]: So this works perfectly fine, but I would actually like to try to standardize on using `"test"` when we can as the `FirstName` or `UserName` of our test users. That way we don't need to think super hard about which test cases use which username.

This isn't documented anywhere, but I would like to strive for all of our tests to be laid out in the same fashion.

File: internal/handlers/telegram/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+			},
+			Caption: "Random Caption",
+		},
+	}
+	// Call the String() function which checks if a username(Optional) exists
+	updateObj.Message.From.UserName = updateObj.Message.From.String()
+
+	// Assign the MimeType(Optional) to FileID(Required) if it does not exists
+	if updateObj.Message.Document.MimeType == ""{
+		updateObj.Message.Document.MimeType = updateObj.Message.Document.FileID
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:27]: We don't want to set the mime type to anything if it doesn't exist. We just skip over it.

File: internal/handlers/telegram/handler_test.go (hunk 1 above)
Comment (Tjzabel) [reply internal/handlers/telegram/handler_test.go:33]: I feel the caption is very useful to users on IRC. Even if there was no caption set, I feel that is still useful information to IRC users.

File: internal/handlers/telegram/handler_test.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+				FirstName: "Someone",
+			},
+			Document: &tgbotapi.Document{
+				FileID: "Random File",
+			},
+			Caption: "Random Caption",
+		},
+	}
+	// Call the String() function which checks if a username(Optional) exists
+	updateObj.Message.From.UserName = updateObj.Message.From.String()
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:23]: Instead of explicitly setting the username to the first name if one doesn't exist, we should just call `user.String()` when we are building the final string to send to IRC.

File: internal/handlers/telegram/handler_test.go (hunk 3 above)
Comment (10eMyrT) [reply internal/handlers/telegram/handler_test.go:14]: You are right. Standardizing would make it easier in the future.

File: internal/handlers/telegram/handler_test.go (hunk 4 above)
Comment (10eMyrT) [reply internal/handlers/telegram/handler_test.go]: I moved it to the `handler.go`.

File: internal/handlers/telegram/handler.go [hunk 6]
```diff
@@ -64,8 +64,12 @@ documentHandler receives a document object from Telegram, and sends
 a notification to IRC.
 */
 func documentHandler(tg *Client, u *tgbotapi.Message) {
-
-	formatted := u.From.UserName + " shared a file (" + u.Document.MimeType + ")"
+	formatted := u.From.String() + " shared a file ("
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:67]: Nice :+1: 
This is exactly how we want to use the `user.String()` method :100:

File: internal/handlers/telegram/handler.go [hunk 7]
```diff
@@ -64,8 +64,12 @@ documentHandler receives a document object from Telegram, and sends
...
+	if u.Document.MimeType != "" {
+		formatted += u.Document.MimeType + ")"
+	} else {
+		formatted += u.Document.FileID + ")"
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:71]: So this is actually the whole issue with #159 
We explicity do _not_ want the unfiltered link to the document download, as this [displays the bot api token](https://github.com/RITlug/teleirc/issues/241). For now, I would recommend only showing the `MimeType` if it is available, and ignore the `FileID` entirely. 

This is something I forgot about in the first review.

File: internal/handlers/telegram/handler_test.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package telegram
+
+import (
+	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestDocument(t *testing.T) {
+	correct := "test shared a file (https://teleirc.com/file.txt) on Telegram with caption: 'Random Caption'."
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:10]: Good, but due to my comment above, we don't actually want to show the file link. 
I would recommend just doing:
```
correct := "test shared a file on Telegram with caption: 'Random Caption'"
```
when there is no mimetype available, as we don't want to show the bot api token.

File: internal/handlers/telegram/handler_test.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	updateObj := &tgbotapi.Update{
+		Message: &tgbotapi.Message{
+			From: &tgbotapi.User{
+				FirstName: "test",
+			},
+			Document: &tgbotapi.Document{
+				FileID: "https://teleirc.com/file.txt",
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:17]: File ID is a required field, so this is good :+1:

File: internal/handlers/telegram/handler.go (hunk 7 above)
Comment (10eMyrT) [reply internal/handlers/telegram/handler.go:71]: I see your point. For security reason, we definitely should not display the api token.

File: internal/handlers/telegram/documentHandler_test.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+/*
+TestDocument_plain checks the behavior of the document handlers if only required
+fields are available.
+*/
+func TestDocument_plain(t *testing.T) {
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler_test.go:13]: I would actually prefer if we try not to use too many underscores in these function names. Me and @Zedjones have only been using underscores to explicitly state when the function booleans are on or off. 
```
func TestDocumentPlain
func TestDocumentCaption
func TestDocumentMime
func TestDocumentUsername
func TestDocumentNoCaption
func TestDocumentFull
```

Comment (10eMyrT) [reply internal/handlers/telegram/documentHandler_test.go:13]: Right, I will solve that problem. Another point, should we all keep our test inside a single file `handlers_test.go` or for each handler, we create a test file?
